test: Test source/sink status reporting

philip-stoev opened the pull request:
A mzcompose-based test that introduces disruptions against a source and a sink and confirms that the disruption is properly reflected in the mz_*_status tables. The disruption is then removed and the test checks that the status tables reflect resumed operation.

### Motivation

  * This PR adds a known-desirable feature.

Relates to #12864

philip-stoev commented on test/source-sink-errors/mzcompose.py:
```diff
@@ -0,0 +1,172 @@
+                  FROM KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-source-topic-${testdrive.seed}')
+                  FORMAT BYTES
+                  ENVELOPE NONE
+                  /* WITH ( REMOTE 'storaged:2100' ) REMOTE causes the status source to be empty */
```
@bkirwi Using the `WITH` clause to connect to a storaged that has been started previously results in an empty status table.

bkirwi replied:
Very puzzling; storaged should behave exactly the same in either case. Perhaps they're being launched with different configurations? I'll look into it.

bkirwi replied:
```suggestion
                  /* TODO(bkirwi) WITH ( REMOTE 'storaged:2100' ) REMOTE causes the status source to be empty */
```

philip-stoev commented on test/source-sink-errors/mzcompose.py:
```diff
@@ -0,0 +1,172 @@
+                  ENVELOPE DEBEZIUM
+                  /* WITH ( REMOTE 'storaged:2100' ) REMOTE causes the status source to be empty */
+
+                # Wait for the source to start up
```
@bkirwi If the disruption happens immediately after `CREATE SOURCE` succeeds, without waiting for the first data to arrive, the status table never transitions away from the `started` state, even though the logs contains the expected `librdkafka` errors.

It is true that in some sense the source never properly got "running", but being stuck in the "started" state and not in "stalled" means that the error column is empty even though librdkafka is already reporting the error in the log.

bkirwi replied:
I agree that `stalled` would be better here... looking into it.

philip-stoev commented on test/source-sink-errors/mzcompose.py:
```diff
@@ -0,0 +1,172 @@
+                  WHERE name = 'source1'
+                stalled true
+
+                #> SELECT status, error ~* '{error}'
```
@bkirwi As discussed previously, sink errors do not appear to be reported.

bkirwi replied:
Working on this in https://github.com/MaterializeInc/materialize/pull/16232

bkirwi replied:
```suggestion
                # TODO(bkirwi) https://github.com/MaterializeInc/materialize/pull/16232
                #> SELECT status, error ~* '{error}'
```

philip-stoev commented on test/source-sink-errors/mzcompose.py:
```diff
@@ -0,0 +1,172 @@
+                > SELECT COUNT(*) FROM source1;
+                2
+
+                #> SELECT status, error
```
@bkirwi After the disruption has been cleared, the source is back up and ingesting , as evidenced by the `SELECT COUNT(*) = 2` check above. However, the status does not always transition to "running".

bkirwi replied:
Investigating this; might be a timing issue.

bkirwi replied:
I think I've figured this one out: an event ordering issue in a multi-worker environment. I suspect most of my local testing was on a single worker.

I'll work up a fix to this, though I'm not sure exactly how to approach it yet.

philip-stoev commented on test/source-sink-errors/mzcompose.py:
```diff
@@ -0,0 +1,172 @@
+        breakage=lambda c, seed: redpanda_topics(c, "delete", seed),
+        expected_error="UnknownTopicOrPartition",
+        fixage=None
+        # Re-creating the topic does not restart the source
```
This is out of scope of this feature, but it seems that re-creating the topic does not cause ingestion to resume

bkirwi replied:
That sounds right, though I'm glad we can at least test that the topic-deleted error percolates up.

bkirwi commented on test/source-sink-errors/mzcompose.py:
```diff
@@ -0,0 +1,172 @@
+                  ENVELOPE DEBEZIUM
+                  /* WITH ( REMOTE 'storaged:2100' ) REMOTE causes the status source to be empty */
+
+                # Wait for the source to start up
```
```suggestion
                # Wait for the source to start up
                # TODO(bkirwi) - catch errors while blocked on client startup
```

bkirwi commented on test/source-sink-errors/mzcompose.py:
```diff
@@ -0,0 +1,172 @@
+                > SELECT COUNT(*) FROM source1;
+                2
+
+                #> SELECT status, error
```
```suggestion
                # TODO(bkirwi) https://github.com/MaterializeInc/materialize/issues/16065
                #> SELECT status, error
```

bkirwi reviewed (approved):
Thank you for this!

I've suggested a bunch of comments that assign future work to me. I propose merging this, and then in future bugfix PRs I can improve the test directly.

philip-stoev commented:
I have incorporated your changes and squashed them in . Will merge the test on green. Please let me know if you need me on tracking any of the residual issues.

I would like to do another round of testing after all the issues have been moved out of the way. In particular, I would also like to do a short cloudtest/K8s test that checks that crash loops are detected properly (if in scope of the task).
